fix: UAA delete user endpoint returns false error during upgrade canary deployment

peterhaochen47 opened the pull request:
- fixes https://github.com/cloudfoundry/uaa/issues/2789 (see bug root cause in the issue)
- by bringing back the MFA-related tables exactly as they were

[#187240345]

Tallicia reviewed (approved):
Restore looks good

peterhaochen47 commented:
Some tests are failing on uploading test results:
```
Run actions/upload-artifact@v4
/usr/bin/docker exec  d78f1e7ca0ca2ef54c3345e3b8c1e9465354a02daefa4d92294c23975801e2f3 sh -c "cat /etc/*release | grep ^ID"
With the provided path, there will be 73 files uploaded
Artifact name is valid!
Root directory input is valid!
Error: Failed to CreateArtifact: Received non-retryable error: Failed request: (409) Conflict: an artifact with this name already exists on the workflow run
```
Not sure what's wrong.

strehle commented on server/src/main/resources/org/cloudfoundry/identity/uaa/db/hsqldb/V4_108__Restore_MFA_Tables.sql:
```diff
@@ -0,0 +1,30 @@
+  config LONGVARCHAR
+);
+
+CREATE UNIQUE INDEX idx_mfa_unique_name ON mfa_providers (identity_zone_id,name);
```
@peterhaochen47 dont know if we should create indexes .
If flyway migrates 76.31.0 -> 77.3.0 then these tables are deleted in a step before , so no content anymore. Then we create the tables again, but wont use it anymore in 77.x. version.
And if 76.31.0 run in parallel, then the empty DB should not be a performance issue

peterhaochen47 replied:
Yes indeed, I was just putting everything in bc I was afraid we would miss something. But no index is fine too.

strehle commented on server/src/main/resources/org/cloudfoundry/identity/uaa/db/postgresql/V4_108__Restore_MFA_Tables.sql:
```diff
@@ -0,0 +1,31 @@
+  config TEXT
+);
+
+CREATE UNIQUE INDEX idx_mfa_unique_name ON mfa_providers (identity_zone_id,LOWER(name));
```
See index comment before

strehle commented:
> Not sure what's wrong.

sql syntax for postgresl/mysql is "CREATE TABLE IF NOT EXISTS", corrected it

peterhaochen47 committed:
fix: UAA delete user endpoint returns false error during upgrade canary deployment

- fixes #2789 (see bug root
  cause in the issue)
- by bringing back the MFA-related tables (but without the index
  as these tables won't actually be used, as the MFA feature itself
  has been removed)

[#187240345]

Co-authored-by: Markus Strehle <11627201+strehle@users.noreply.github.com>

strehle commented on server/src/main/resources/org/cloudfoundry/identity/uaa/db/postgresql/V4_108__Restore_MFA_Tables.sql:
```diff
@@ -0,0 +1,29 @@
+--
```
4_108 is now used from recent merge, so if you want this please incr. number

strehle reviewed (commented):
Since you have not merged it yesterday for the release I dont see a reason now to have it, correct ?

peterhaochen47 commented:
> Since you have not merged it yesterday for the release I dont see a reason now to have it, correct ?

We have not merged it because this is still under discussion in slack, and this PR has not received enough approvals.

strehle commented:
> > Since you have not merged it yesterday for the release I dont see a reason now to have it, correct ?
> 
> We have not merged it because this is still under discussion in slack, and this PR has not received enough approvals.

ok, so you would get an approval, but my personal opinion is, now it is too late and I would not restore tables. UAA was not crashed,.... So too much effort for too less impact

strehle reviewed (approved):
please take care for correct versioning of flyway file ... 108 is used alredy.

my personal opinion is, now it is too late and I would not restore tables. UAA was not crashed,.... So too much effort for too less impact

However you can have an approval

peterhaochen47 commented:
Let's keep this PR as a draft, and wait until next week, then I'll discuss this with @bruce-ricard, who worked on this issue with me.

peterhaochen47 commented:
I have discussed with @bruce-ricard, and decided that if SAP, the party that reported this issue, does not care about this fix, and given that our side has not received any reports of this issue, we can close this PR for now. If there's any new complaints, we can always reopen this PR.

For now, we have documented this issue in release notes and this open issue: https://github.com/cloudfoundry/uaa/issues/2789 And we see that as sufficient for now.
